Content.Server/GameObjects/Components/Power/PowerCellSlotComponent.cs

```diff
@@ -0,0 +1,244 @@
+        /// <summary>
+        /// True if we don't want a cell inserted during map init.
+        /// </summary>
+        private bool _startEmpty = false;
+
+        /// <summary>
+        /// If not null, this cell type will be inserted at MapInit instead of the default Standard cell.
+        /// </summary>
+        private string? _startingCellType = null;
+
+        public override void ExposeData(ObjectSerializer serializer)
+        {
+            base.ExposeData(serializer);
+            serializer.DataField(ref SlotSize, "slotSize", PowerCellSize.Small);
+            serializer.DataField(ref CanRemoveCell, "canRemoveCell", true);
+            serializer.DataField(ref ShowVerb, "showVerb", true);
+            serializer.DataField(ref _startEmpty, "startEmpty", false);
+            serializer.DataField(ref _startingCellType, "startingCellType", null);
+            serializer.DataField(ref CellRemoveSound, "cellRemoveSound", "/Audio/Items/pistol_magin.ogg");
+            serializer.DataField(ref CellInsertSound, "cellInsertSound", "/Audio/Items/pistol_magout.ogg");
+            serializer.DataField(ref DescFormatString, "descFormatString", "It uses size {0} power cells.");
+        }
+
+        public override void Initialize()
+        {
+            base.Initialize();
+            _cellContainer = ContainerManagerComponent.Ensure<ContainerSlot>("cellslot_cell_container", Owner, out _);
+        }
+
+        void IExamine.Examine(FormattedMessage message, bool inDetailsRange)
+        {
+            string sizeLetter = SlotSize switch
+            {
+                PowerCellSize.Small => Loc.GetString("S"),
+                PowerCellSize.Medium => Loc.GetString("M"),
+                PowerCellSize.Large => Loc.GetString("L"),
+                _ => "???"
+            };
+            if (DescFormatString != null) message.AddMarkup(string.Format(DescFormatString, sizeLetter));
+        }
+
+        /// <summary>
+        /// Remove the cell from this component. If a user is specified, the cell will be put in their hands
+        /// or failing that, at their feet. If no user is specified the cell will be put at the location of
+        /// the parent of this component.
+        /// </summary>
+        /// <param name="user">(optional) the user to give the removed cell to.</param>
+        /// <param name="playSound">Should <see cref="CellRemoveSound"/> be played upon removal?</param>
+        /// <returns>The cell component of the entity that was removed, or null if removal failed.</returns>
+        public PowerCellComponent? EjectCell(IEntity? user, bool playSound = true)
+        {
+            var cell = Cell;
+            if (cell == null || !CanRemoveCell) return null;
+            if (!_cellContainer.Remove(cell.Owner)) return null;
+            //Dirty();
+            if (user != null)
+            {
+                if (!user.TryGetComponent(out HandsComponent? hands) || !hands.PutInHand(cell.Owner.GetComponent<ItemComponent>()))
```

Comment:
You check for ItemComponent on ejecting but don't check upon inserting so this could potentially throw.